File: src/main/java/de/twometer/orion/res/ShaderLoader.java
package de.twometer.orion.res;

import de.twometer.orion.util.Log;

import java.io.IOException;

import static org.lwjgl.opengl.GL20.*;
import static org.lwjgl.opengl.GL20.glDeleteShader;

public class ShaderLoader {

    public static int loadShader(String vert, String frag) {
        Log.d("Loading shader " + vert + "; " + frag);

        String vertPath = AssetPaths.SHADER_PATH + vert;
        String fragPath = AssetPaths.SHADER_PATH + frag;

        try {

            int vertexShaderId = glCreateShader(GL_VERTEX_SHADER);
            int fragmentShaderId = glCreateShader(GL_FRAGMENT_SHADER);

            String vertexShaderCode = ResourceLoader.loadString(vertPath);
            String fragmentShaderCode = ResourceLoader.loadString(fragPath);

            glShaderSource(vertexShaderId, vertexShaderCode);
            glCompileShader(vertexShaderId);
            checkShaderError(vertexShaderId);

            glShaderSource(fragmentShaderId, fragmentShaderCode);
            glCompileShader(fragmentShaderId);
            checkShaderError(fragmentShaderId);

            int programId = glCreateProgram();
            glAttachShader(programId, vertexShaderId);
            glAttachShader(programId, fragmentShaderId);
            glLinkProgram(programId);

            // After the shader program is linked, the shader sources can be cleaned up
            glDetachShader(programId, vertexShaderId);
            glDetachShader(programId, fragmentShaderId);
            glDeleteShader(vertexShaderId);
            glDeleteShader(fragmentShaderId);

            return programId;
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private static void checkShaderError(int shaderId) {
        String log = glGetShaderInfoLog(shaderId);
        if (log.length() > 0) {
            Log.e("Shader compilation failed");
            Log.e(log);
        }
    }

}

File: neko-engine/src/main/kotlin/de/twometer/neko/res/ShaderLoader.kt
package de.twometer.neko.res

import de.twometer.neko.render.Shader
import de.twometer.neko.render.ShaderInject
import de.twometer.neko.render.ShaderProperty
import de.twometer.neko.res.ShaderParser.SECTION_FRAGMENT
import de.twometer.neko.res.ShaderParser.SECTION_SHARED
import de.twometer.neko.res.ShaderParser.SECTION_VERTEX
import mu.KotlinLogging
import org.lwjgl.opengl.GL20.*

private val logger = KotlinLogging.logger {}

object ShaderLoader {

    fun load(path: String): Shader {
        logger.info { "Loading nks shader $path" }

        val ast = ShaderParser.loadShaderAst(path)

        var vertexShader = ast.getSectionContent(SECTION_VERTEX)
        var fragmentShader = ast.getSectionContent(SECTION_FRAGMENT)

        when {
            vertexShader == null -> failure("Vertex shader not defined")
            fragmentShader == null -> failure("Fragment shader not defined")
        }

        ast.sections[SECTION_SHARED]?.forEach {
            if (it is ShaderParser.PlainText) {
                vertexShader = "out ${it.data}\n$vertexShader"
                fragmentShader = "in ${it.data}\n$fragmentShader"
            }
        }

        vertexShader = "#version ${ast.version}\n$vertexShader"
        fragmentShader = "#version ${ast.version}\n$fragmentShader"

        val shader = Shader(loadGlsl(vertexShader!!, fragmentShader!!))

        shader.bind()
        ast.directives.forEach {
            when (it.type) {
                ShaderParser.DirectiveType.Bind -> {
                    shader[it.arguments[0]] = it.arguments[1].toInt()
                }
                ShaderParser.DirectiveType.Inject -> {
                    shader.injects.add(ShaderInject.valueOf(it.arguments.first()))
                }
                ShaderParser.DirectiveType.Set -> {
                    shader.properties[ShaderProperty.valueOf(it.arguments[0])] = it.arguments[1]
                }
                else -> {
                }
            }
        }
        shader.unbind()

        return shader
    }

    private fun loadGlsl(vertexSrc: String, fragmentSrc: String): Int {
        val vertexShaderId = glCreateShader(GL_VERTEX_SHADER)
        val fragmentShaderId = glCreateShader(GL_FRAGMENT_SHADER)

        glShaderSource(vertexShaderId, vertexSrc)
        glCompileShader(vertexShaderId)
        checkShaderErrors(vertexShaderId, vertexSrc)

        glShaderSource(fragmentShaderId, fragmentSrc)
        glCompileShader(fragmentShaderId)
        checkShaderErrors(fragmentShaderId, fragmentSrc)

        val programId = glCreateProgram()
        glAttachShader(programId, vertexShaderId)
        glAttachShader(programId, fragmentShaderId)
        glLinkProgram(programId)

        // After the shader program is linked, the shader sources can be cleaned up
        glDetachShader(programId, vertexShaderId)
        glDetachShader(programId, fragmentShaderId)
        glDeleteShader(vertexShaderId)
        glDeleteShader(fragmentShaderId)

        return programId
    }

    private fun checkShaderErrors(shaderId: Int, src: String) {
        val log = glGetShaderInfoLog(shaderId)
        if (log.isNotEmpty()) {
            val logLines = log.split("\n")

            logger.error { "Shader compilation failed" }
            logLines.forEach {
                if (it.isNotBlank() && it.contains("error"))
                    logger.error { reformatLogLine(it, src) }
            }

            failure("Could not compile GLSL")
        }
    }

    private fun reformatLogLine(logLine: String, src: String): String {
        val glslLine = logLine.substringAfter("(").substringBefore(")").toInt()
        val glslErrorMessage = logLine.substringAfter(':').trim()

        val nksRefInfo = src.split("\n")[glslLine - 1].substringAfterLast("REF:").split(":")
        val nksFileName = nksRefInfo[0]
        val nksLineNo = nksRefInfo[1]

        return "[$nksFileName:$nksLineNo] $glslErrorMessage"
    }

    private fun failure(message: String): Nothing {
        error("Shader compilation error: $message")
    }

}